src/userdev/java/net/minecraftforge/gradle/userdev/MinecraftUserRepo.java

```diff
@@ -176,6 +207,64 @@ public void validate(Configuration cfg, Map<String, RunConfig> runs, File native
                 runs.computeIfAbsent(name, k -> new RunConfig()).merge(dev, false, vars);
             });
         }
+        this.extraDataFiles = this.buildExtraDataFiles();
+    }
+
+    /**
+     * Previously, Configuration.resolve() was called indirectly from
+     * BaseRepo.getArtifact(). Due to the extensive amount of Gradle code that
+     * runs as a result of the resolve() call, deadlock could occur as follows:
+     *
+     * 1. Thread #1: During resolution of a dependency, Gradle calls
+     * BaseRepo#getArtifact(). The 'synchronized' block is entering,
+     * causing a lock to be taken on the artifact name
+     *
+     * 2. Thread #2: On a different thread, internal Gradle code takes a lock
+     * in the class org.gradle.internal.event.DefaultListenerManager.EventBroadcast.ListenerDispatch
+     *
+     * 3. Thread #1: Execution continues on the 'BaseRepo#getArtifact()' call
+     * stack, reaching the call to Configuration.resolve(). This call leads
+     * to Gradle internally dispatching events through the same class
+     * org.gradle.internal.event.DefaultListenerManager.EventBroadcast.ListenerDispatch.
+     * This thread is now blocked on the internal Gradle lock taken by Thread #2
+     *
+     * 4. Thread #2: Execution continues, and attempts to resolve the same
+     * dependency that Thread #1 is currently resolving. Since Thread #1 is
+     * still in the 'synchronized' block with the same artifact name, Thread #2
+     * blocks.
+     *
+     * These threads are now deadlocked: Thread #1 is holding the
+     * BaseRepo#getArtifact lock while waiting on an internal Gradle lock,
+     * while Thread #2 is holding the same internal Gradle lock while waiting
+     * on the BaseRepo#getArtifact lock.
+     *
+     * Visit https://git.io/fhHLk to see a stack dump showing this deadlock
+     *
+     * Fortunately, the solution is fairly simply. We can move the entire
+     * dependency creation/resolution block to an earlier point in
+     * ForgeGradle's execution. Since the client 'data'/'extra'/libraries only
+     * depend on the MCP config file, we can do this during plugin
+     * initialization.
+     *
+     * This has the added benefit of speeding up ForgeGradle - this block of
+     * code will only be executed once, instead of during every call to
+     * compileJava
+     * @return
+     */
+    private Set<File> buildExtraDataFiles() {
+        Configuration cfg = project.getConfigurations().create(getNextCompileName());
+        List<String> deps = new ArrayList<>();
+        deps.add("net.minecraft:client:" + mcp.getMCVersion() + ":extra");
+        deps.add("net.minecraft:client:" + mcp.getMCVersion() + ":data");
+        deps.addAll(mcp.getLibraries());
```

Comment:
Just as a note, as I was the one to originally hardcode 'client'. It is it necessary to use the client, or should this be moved to the proper side as well? So Server only stuff gets the proper libraries.

Also, speaking with cpw, there is a issue with dedicated server and log4j with extra/data being separate {Log4j tries to grab the custom logger class from the same jar as the config, instead of from the classpath}. So it's on my TODO to merge the 'extra' and 'data' deps into just a single 'extra' jar.